File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans= INT_MIN;
    int h(TreeNode* root){
        if(root == NULL) return 0;
        int lh = max(0,h(root->left));
        int rh = max(0,h(root->right));
        ans = max(ans,root->val+lh+rh);
        return root->val + max(lh,rh);
    }
    int maxPathSum(TreeNode* root) {
        h(root);
        return ans;
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxi=Integer.MIN_VALUE;
    public int h(TreeNode root){
        if(root == null) return 0;
        int lh= Math.max(0,h(root.left));
        int rh= Math.max(0,h(root.right));
        maxi= Math.max(maxi,root.val+lh+rh);
        return root.val + Math.max(lh,rh);
    }
    public int maxPathSum(TreeNode root) {
        h(root);
        return maxi;
    }
}